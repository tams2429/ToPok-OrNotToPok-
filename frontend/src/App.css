*
{
    -webkit-box-sizing: border-box;
       -moz-box-sizing: border-box;
            box-sizing: border-box;
}

.App {
  text-align: center;
  background-color: #323D48;
  min-height: 100vh;
}

.App-header {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: 20vmin;
  color: #ffd52d;
  font-family: 'Pokemon Hollow', sans-serif;
  padding: 3vmin;
}

.loading-wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
}

.pokeball {
  width: 60px;
  height: 60px;
  background-color: #fff;
  border-radius: 50% ;
  position: relative;
  overflow: hidden;
  border: 3px solid;
  animation: frames .8s  linear 0s infinite;
}

.pokeball:after{
  content: '';
  position: absolute;
  width: 60px;
  height: 60px;
  background-color: red;
  border-bottom: 4px solid;
  top: -4px
}

.pokeball:before{
  content: '';
  position: absolute;
  background-color: #fff;
  width: 10px;
  height: 10px;
  border: 4px solid;
  border-radius: 50%;
  bottom: 18px;
  right: 18px;
  z-index: 1;
}

.classic-1 {
  font-weight: bold;
  font-family: monospace;
  font-size: 30px;
  display: grid;
}

.classic-1:before,
.classic-1:after {
  content:"Loading...";
  grid-area: 1/1;
  -webkit-mask:linear-gradient(90deg,#000 50%,#0000 0) 0 50%/2ch 100%;
  animation:c1 1s infinite cubic-bezier(0.5,220,0.5,-220);
}
.classic-1:after {
  -webkit-mask-position:1ch 50%;
}

/* AnimationFrames */
@keyframes c1 {100%{transform: translateY(calc(var(--s,1)*0.1%));}}
@keyframes frames{
  0% {
    transform: rotate(0deg);
  }
  100%{
    transform: rotate(360deg);
  }
}
