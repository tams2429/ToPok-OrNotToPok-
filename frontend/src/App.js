import axios from 'axios'

// COMPONENTS
import { useState, useEffect } from 'react'
import { SearchBar } from './components/SearchBar'
import { SearchBtn } from './components/SearchBtn'
import { PokemonEntry } from './components/PokemonEntry'

// STYLESHEETS
import './App.css';

export const App = () => {

  const [searchString, setSearchString] = useState('')
  const [loading, setLoading] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')
  const [pokemonInfo, setPokemonInfo] = useState(null)

  console.log({searchString})

  //TODO: refactor by extracting constants to a constants file
  const pokemonSearchEndpointPrefix = '/pokemon/'
  const totalSearchablePokemon = 898

  // Upon completing initial render, enforced an initial 2 second delay to reset loading state back to false for aesthetic purposes
  // Can remove to improve perfomance
  // useEffect(() => {
  //   const timeout = setTimeout(() => setLoading(false), 2000);
  //   return () => clearTimeout(timeout);
  // }, []);

  //API call/function to extract logic later
  const handleSearch = async (e) => {
    // console.log('form submitted')
    e.preventDefault()
    //Clear any existing errorMessage
    setErrorMessage('')

    //Logic to convert searchString to lowercase
    setLoading(true)
    const lowerCaseSearchString = searchString.toLowerCase()
    //TODO: refactor by extracting following logic into a function and invoke
    try {
      
      const response = await axios.get(pokemonSearchEndpointPrefix + lowerCaseSearchString)
      console.log('response is:', response)
      setPokemonInfo(response.data)
      setLoading(false)
    } 
    catch (err) {
      console.log('error is:', err)
      setLoading(false)
      setErrorMessage(err.response.data.errorMessage)
    }
  }

  // Random pokemon id upto and including id: 898
  const handleRandomSearch = async() => {
    const randomId = Math.floor(Math.random() * totalSearchablePokemon) + 1
    console.log({randomId})

    setErrorMessage('')
    setLoading(true)
    //TODO: refactor by extracting following logic into a function and invoke
    try {
      const response = await axios.get(pokemonSearchEndpointPrefix + randomId.toString())
      setPokemonInfo(response.data)
      setLoading(false)
    } 
    catch (err) {
      setLoading(false)
      setErrorMessage(err.response.data.errorMessage)
    }
  }

  const mockResponse = {
    "name": "mewtwo",
    "description": "'t wast did create by a scientist after years of horrific gene splicing and dna engineering experiments.",
    "sprite": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
  }

  console.log('Pokemon info is:', pokemonInfo)
  console.log('Error message is:', errorMessage)
  return (
    <div className='App'>
      <header className='App-header'>
        Pokéspear
      </header>
      {loading ?
      //TODO: refactor by extracting into a loading component
      <div className='loading-wrapper'>
        <p className='classic-1' data-testid='loading-text'/>
        <div className='pokeball' />
      </div>
       :
        //TODO: refactor by extracting to a separate SearchForm component?
        //DONE => removed onClick={handleSearch} from <SearchBtn /> props and type=submit => to test if it works?
      <>
        <form onSubmit={handleSearch}>
          <SearchBar searchString={searchString} setSearchString={setSearchString} handleSearch={handleSearch}/>
          <SearchBtn btnText='Poké Search' btnType='submit' searchString={searchString} isDisabled={!searchString}/>
          <SearchBtn btnText='Surprise Me' btnType='button' isDisabled={false} onClick={handleRandomSearch}/>
        </form>  
        {errorMessage ? 
        <h2 className='error-message'>{errorMessage}</h2>
        :
        pokemonInfo && <PokemonEntry {...pokemonInfo}/>
        // mockResponse && <PokemonEntry {...mockResponse}/>
        }
        {/* {mockResponse && <PokemonEntry {...mockResponse}/>} */}
      </>
      }
    </div>

  );
}
