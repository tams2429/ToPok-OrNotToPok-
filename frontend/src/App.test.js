import { render, screen } from '@testing-library/react';
import App from './App';

test('renders learn react link', () => {
  render(<App />);
  const linkElement = screen.getByText(/learn react/i);
  expect(linkElement).toBeInTheDocument();
});

const mockResponse = {
  "name": "mewtwo",
  "description": "'t wast did create by a scientist after years of horrific gene splicing and dna engineering experiments.",
  "sprite": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
}



// Test that the correct components have been initally rendered against snapshot (before API interaction)

// Test that loading state components have been rendered when loading state is true

// Test that the MockAPI is called when user interacts with the PokeSearch btn + with the correct parameters (i.e. with the MockPokemonName)

// Test that the MockAPI is called when the user clicks on the SurpriseMe btn correctly with the correct parameters + endpoint

// Test that the correct components have been rendered against snapshot after receiving MockPokemonInfo (i.e. response from API)
