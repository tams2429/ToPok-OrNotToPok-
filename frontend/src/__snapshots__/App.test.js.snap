// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main application should fetch and display pokemon info when user clicks on the Surprise Me button (i.e. random search) 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      Pokéspear
    </header>
    <form>
      <div>
        <input
          aria-label="Searchbar-for-pokemon"
          class="searchBar"
          data-testid="searchBar"
          placeholder="Search your Pokémon here..."
          type="search"
          value=""
        />
      </div>
      <button
        class="searchBtn disabled"
        disabled=""
        type="submit"
      >
        Poké Search
      </button>
      <button
        class="searchBtn "
        type="button"
      >
        Surprise Me
      </button>
    </form>
    <section
      class="pokemon-entry-container"
    >
      <img
        alt="Searched pokemon sprite appearance"
        class="pokemon-img"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
      />
      <h2
        class="pokemon-name"
      >
        mewtwo
      </h2>
      <p
        class="pokemon-description"
      >
        't wast did create by a scientist after years of horrific gene splicing and dna engineering experiments.
      </p>
    </section>
  </div>
</div>
`;

exports[`Main application should fetch and display the pokemon info when the user searches for a valid pokemon 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      Pokéspear
    </header>
    <form>
      <div>
        <input
          aria-label="Searchbar-for-pokemon"
          class="searchBar"
          data-testid="searchBar"
          placeholder="Search your Pokémon here..."
          type="search"
          value="mewtwo"
        />
      </div>
      <button
        class="searchBtn "
        type="submit"
      >
        Poké Search
      </button>
      <button
        class="searchBtn "
        type="button"
      >
        Surprise Me
      </button>
    </form>
    <section
      class="pokemon-entry-container"
    >
      <img
        alt="Searched pokemon sprite appearance"
        class="pokemon-img"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
      />
      <h2
        class="pokemon-name"
      >
        mewtwo
      </h2>
      <p
        class="pokemon-description"
      >
        't wast did create by a scientist after years of horrific gene splicing and dna engineering experiments.
      </p>
    </section>
  </div>
</div>
`;

exports[`Main application should render properly before search and match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      Pokéspear
    </header>
    <form>
      <div>
        <input
          aria-label="Searchbar-for-pokemon"
          class="searchBar"
          data-testid="searchBar"
          placeholder="Search your Pokémon here..."
          type="search"
          value=""
        />
      </div>
      <button
        class="searchBtn disabled"
        disabled=""
        type="submit"
      >
        Poké Search
      </button>
      <button
        class="searchBtn "
        type="button"
      >
        Surprise Me
      </button>
    </form>
  </div>
</div>
`;

exports[`Main application should render the loading components when in a loading state 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      Pokéspear
    </header>
    <div
      class="loading-wrapper"
    >
      <p
        class="classic-1"
        data-testid="loading-text"
      />
      <div
        class="pokeball"
      />
    </div>
  </div>
</div>
`;

exports[`Main application should throw an error message when the user searches for an invalid pokemon/error with network 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      Pokéspear
    </header>
    <form>
      <div>
        <input
          aria-label="Searchbar-for-pokemon"
          class="searchBar"
          data-testid="searchBar"
          placeholder="Search your Pokémon here..."
          type="search"
          value="qwerty"
        />
      </div>
      <button
        class="searchBtn "
        type="submit"
      >
        Poké Search
      </button>
      <button
        class="searchBtn "
        type="button"
      >
        Surprise Me
      </button>
    </form>
    <h2
      class="error-message"
    >
      400: Looks like something went wrong, please try again
    </h2>
  </div>
</div>
`;
